### DISCUSSION TITLE
psql-http / Webhooks transaction issue

### DISCUSSION THREAD

**Discussion by quicksnap** [supabase/supabase#1594] [Questions]
This is sort of a continuation on https://github.com/supabase/supabase/discussions/259.

I'm trying to set up push notifications such that after an insert, I hit a lamda function. I figured I would use `psql-http`, but I ran into a snag: 

- `psql-http` runs within the transaction
- calling my lamda endpoint via `http_get(...)`, I'm wanting to collect the information by simply having the lamda fn use the supabase API. I'd rather not send all the data in the webhook request
- But, because `pgsql-http` is within the transaction, data isn't committed yet, because it's waiting for an HTTP response!
- If I send a response, my lamda function is killed, so I can't do the work.

Anyone know any tricks I could use to get around this? `psql-http` does not yet support background workers. I'm looking to avoid deploying an actual server. I really just want Supabase to fire-and-forget a request to my endpoint.

**Comment by quicksnap** [comment]
Update: Some helpful people clued me into a workaround:

- Call a serverless endpoint
- Have said endpoint call _another endpoint_, but make sure it fully sends the request before sending a response
- Have second endpoint `sleep()` for a bit while the DB commits.

Relevant links:
- https://www.sensedeep.com/blog/posts/stories/lambda-fast-http.html
- http://yurigor.com/how-to-make-http-request-from-node-js-without-waiting-for-response/

**Comment by inian** [comment]
Can you post the code for calling the lambda endpoint via `psql-http` @quicksnap? This is something I wanted to try out too 😄  
This is a perfect use case for our workflows engine which we are going to launch in a few months time. 

In the mean time, another way to do this would be to launch [a step function](https://aws.amazon.com/step-functions/) via psql-http. You can use the [Wait state](https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-wait-state.html) to sleep for a second or two and then make it trigger your actual lambda function

**Reply by quicksnap** [reply]
I'm _actually_ using Vercel functions, so I don't get to use any AWS stuff directly. I'm trying to keep my setup as easy-mode as possible, since it's just me on this project of mine.

Trigger code here. Some of this stuff, like `VOLATILE NOT LEAKPROOF SECURITY DEFINER` was generated by some tool somewhere.. but it seemed OK enough. Postgres isn't my strong suit.

```sql
CREATE FUNCTION public.notif_webhook_trigger()
    RETURNS trigger
    LANGUAGE 'plpgsql'
    COST 100
    VOLATILE NOT LEAKPROOF SECURITY DEFINER
AS $BODY$
begin
  perform * from http_get('myEndpoint');
  return new;
end;
$BODY$;

CREATE TRIGGER on_new_notification_created
    AFTER INSERT
    ON public.notifications
    FOR EACH STATEMENT
    EXECUTE FUNCTION public.notif_webhook_trigger();
```
